server/data/data-jpa/src/main/java/org/apache/james/rrt/jpa/JPARecipientRewriteTable.java

```diff
@@ -18,204 +18,29 @@
  ****************************************************************/
 package org.apache.james.rrt.jpa;
 
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-
 import javax.inject.Inject;
-import javax.persistence.EntityManager;
 import javax.persistence.EntityManagerFactory;
-import javax.persistence.EntityTransaction;
-import javax.persistence.PersistenceException;
 import javax.persistence.PersistenceUnit;
 
-import org.apache.james.backends.jpa.EntityManagerUtils;
-import org.apache.james.core.Domain;
-import org.apache.james.rrt.api.RecipientRewriteTableException;
-import org.apache.james.rrt.jpa.model.JPARecipientRewrite;
-import org.apache.james.rrt.lib.AbstractRecipientRewriteTable;
-import org.apache.james.rrt.lib.Mapping;
-import org.apache.james.rrt.lib.MappingSource;
-import org.apache.james.rrt.lib.Mappings;
-import org.apache.james.rrt.lib.MappingsImpl;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import org.apache.james.domainlist.api.DomainList;
+import org.apache.james.rrt.lib.RecipientRewriteTableImpl;
 
 /**
  * Class responsible to implement the Virtual User Table in database with JPA
  * access.
  */
-public class JPARecipientRewriteTable extends AbstractRecipientRewriteTable {
-    private static final Logger LOGGER = LoggerFactory.getLogger(JPARecipientRewriteTable.class);
-
-    /**
-     * The entity manager to access the database.
-     */
-    private EntityManagerFactory entityManagerFactory;
+public class JPARecipientRewriteTable extends RecipientRewriteTableImpl<JPARecipientRewriteTableDAO> {
```

Comment:
Here you redefine an `Impl` while, all the specific parts should be in `*DAO`.

That's why there is not point of having an interface for one class.
The only hard-coded behavior is how you use your DAO, which is what you want.

Reply:
The point is tricking Spring.

We keep the same FQDN for people to write down in their domainlist.xml configuration file. 

This `workaround` enables not rewriting Spring configuration. That's not in my plans, either short term or long term.

The design is what you describe for everything that don't need to be defined with Spring (basically just Cassandra), but for everything that needs to be specified with Spring:
 - Write a DAO
 - Reuser XXXImpl
 - Expose a class extending Impl with the same FQDN than before for API stability

Reply:
Impl have dependencies not present in `data-api`, I'm reluctant to add them there.

Also, If I wish to write a RRT implementation that do zero rewrites or whatever please me I need to have an API from it.